pkg/metrics/statsd/statsd_test.go

```diff
@@ -83,3 +85,240 @@ func TestEmitFloat(t *testing.T) {
 		t.Error(m)
 	}
 }
+
+func TestParseSocketEnv(t *testing.T) {
+	for _, tt := range []struct {
+		name       string
+		teststring string
+		part1      string
+		part2      string
+		wantError  bool
+	}{
+		{
+			name:       "Valid string",
+			teststring: "part1:part2",
+			part1:      "part1",
+			part2:      "part2",
+			wantError:  false,
+		},
+		{
+			name:       "Empty first part-regarded valid",
+			teststring: ":part2",
+			part1:      "",
+			part2:      "part2",
+			wantError:  false,
+		},
+		{
+			name:       "Empty second part-regarded valid",
+			teststring: "part1:",
+			part1:      "part1",
+			part2:      "",
+			wantError:  false,
+		},
+		{
+			name:       "No separator",
+			teststring: "somerandommtext",
+			wantError:  true,
+		},
+		{
+			name:       "Empty string",
+			teststring: "",
+			wantError:  true,
+		},
+		{
+			name:       "More than one separator",
+			teststring: "a:b:c",
+			part1:      "a",
+			part2:      "b:c",
+			wantError:  false,
+		},
+		{
+			name:       "Many separators",
+			teststring: ":::::",
+			part1:      "",
+			part2:      "::::",
+			wantError:  false,
+		},
+		{
+			name:       "One separator only",
+			teststring: ":",
+			part1:      "",
+			part2:      "",
+			wantError:  false,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &statsd{}
+			part1, part2, err := s.parseSocketEnv(tt.teststring)
+
+			if tt.wantError {
+				if err == nil {
+					t.Error(fmt.Errorf("expected error but didn't get one."))
+				}
+			} else {
+				if err != nil {
+					t.Error(fmt.Errorf("unexpected error received: %q ", err.Error()))
+				}
+				if part1 != tt.part1 {
+					t.Error(fmt.Errorf("part1 does not match: Wanted: %q but got %q ", tt.part1, part1))
+				}
+				if part2 != tt.part2 {
+					t.Error(fmt.Errorf("part2 does not match: Wanted: %q but got %q ", tt.part2, part2))
+				}
+			}
```

Comment:
1. I would drop error formatting. `t.Error(actualValue)` will print test file name and line number as well as name of the test case from `t.Run(...)`). This will be enough to indetify what when what's wrong.
2. IMO depsite the fact that we no longer return multiple errors from the function I still think that `tt.wantError` should be a string as we have multiple different cases where we hit an error and we might add more erorrs later into this function. Will be easier to extend the test.
3. IMO, `part1` and `part2` should be tested independently from an error. This will help us to ensure that when we return an error we return an empty parts
4. `part1` & `part2`/`tt.part1` & `tt.part2` need better name. Maybe `network` & `address` / `tt.wantNetwork` & `tt.wantAddress`?

Reply:
This is what I came up with while reviewing, if you are interested. This is the above feedback applied to all the new tests in `pkg/metrics/statsd/statsd_test.go` file.

<details>
<summary>Spoiler</summary>


```go
func TestParseSocketEnv(t *testing.T) {
	for _, tt := range []struct {
		name        string
		teststring  string
		wantNetwork string
		wantAddress string
		wantError   string
	}{
		{
			name:        "Valid string",
			teststring:  "part1:part2",
			wantNetwork: "part1",
			wantAddress: "part2",
		},
		{
			name:        "Empty first part-regarded valid",
			teststring:  ":part2",
			wantNetwork: "",
			wantAddress: "part2",
		},
		{
			name:        "Empty second part-regarded valid",
			teststring:  "part1:",
			wantNetwork: "part1",
			wantAddress: "",
		},
		{
			name:       "No separator",
			teststring: "somerandommtext",
			wantError:  "malformed definition for the mdm statds socket. Expecting udp:<hostname>:<port> or unix:<path-to-socket> format. Got: \"somerandommtext\"",
		},
		{
			name:       "Empty string",
			teststring: "",
			wantError:  "malformed definition for the mdm statds socket. Expecting udp:<hostname>:<port> or unix:<path-to-socket> format. Got: \"\"",
		},
		{
			name:        "More than one separator",
			teststring:  "a:b:c",
			wantNetwork: "a",
			wantAddress: "b:c",
		},
		{
			name:        "Many separators",
			teststring:  ":::::",
			wantNetwork: "",
			wantAddress: "::::",
		},
		{
			name:        "One separator only",
			teststring:  ":",
			wantNetwork: "",
			wantAddress: "",
		},
	} {
		t.Run(tt.name, func(t *testing.T) {
			s := &statsd{}
			network, address, err := s.parseSocketEnv(tt.teststring)

			if err != nil && err.Error() != tt.wantError ||
				err == nil && tt.wantError != "" {
				t.Fatal(err)
			}

			if network != tt.wantNetwork {
				t.Error(network)
			}
			if address != tt.wantAddress {
				t.Error(address)
			}
		})
	}
}

func TestValidateSocketDefinition(t *testing.T) {
	for _, tt := range []struct {
		name         string
		network      string
		address      string
		expectToPass bool
		wantError    string
	}{
		{
			name:         "Valid UDP case",
			network:      "udp",
			address:      "127.0.0.1:9000",
			expectToPass: true,
		},
		{
			name:         "Valid Unix Domain Socket case",
			network:      "unix",
			address:      "/var/something/or/another",
			expectToPass: true,
		},
		{
			name:         "Invalid protocoll",
			network:      "tcp",
			address:      "127.0.0.1:12",
			expectToPass: false,
			wantError:    "unsupported protocol for the mdm statds socket. Expecting  'udp:' or 'unix:'. Got: \"tcp\"",
		},
		{
			name:         "Valid protocol, random invalid address - is good here",
			network:      "udp",
			address:      "somerandomtext149203#$%",
			expectToPass: true,
		},
	} {
		t.Run(tt.name, func(t *testing.T) {
			s := &statsd{}

			ok, err := s.validateSocketDefinition(tt.network, tt.address)

			if err != nil && err.Error() != tt.wantError ||
				err == nil && tt.wantError != "" {
				t.Fatal(err)
			}

			if ok != tt.expectToPass {
				t.Error(ok)
			}
		})
	}
}

func TestConnectionDetails(t *testing.T) {
	for _, tt := range []struct {
		name            string
		isLocalDev      bool
		mdmsocketstring string
		network         string
		address         string
		wantError       string
	}{
		{
			name:            "Old behaviour / production mode",
			isLocalDev:      false,
			mdmsocketstring: "",
			network:         "unix",
			address:         "/var/etw/mdm_statsd.socket",
			wantError:       "",
		},
		{
			name:            "Old Behaviour / localdev mode",
			isLocalDev:      true,
			mdmsocketstring: "",
			network:         "unix",
			address:         "mdm_statsd.socket",
			wantError:       "",
		},
		{
			name:            "Valid UDP variable",
			isLocalDev:      false,
			mdmsocketstring: "udp:127.0.0.1:9000",
			network:         "udp",
			address:         "127.0.0.1:9000",
			wantError:       "",
		},
		{
			name:            "Don't override default in localdev ",
			isLocalDev:      true,
			mdmsocketstring: "udp:127.0.0.1:9000",
			network:         "udp",
			address:         "127.0.0.1:9000",
			wantError:       "",
		},
		{
			name:            "Random string without separator",
			isLocalDev:      true,
			mdmsocketstring: "another random string without separator",
			wantError:       "malformed definition for the mdm statds socket. Expecting udp:<hostname>:<port> or unix:<path-to-socket> format. Got: \"another random string without separator\"",
		},
		{
			name:            "Invalid Protocol",
			isLocalDev:      true,
			mdmsocketstring: "foo:bar",
			wantError:       "unsupported protocol for the mdm statds socket. Expecting  'udp:' or 'unix:'. Got: \"foo\"",
		},
	} {
		t.Run(tt.name, func(t *testing.T) {
			controller := gomock.NewController(t)
			defer controller.Finish()
			env := mock_env.NewMockInterface(controller)

			if tt.mdmsocketstring == "" {
				if tt.isLocalDev {
					env.EXPECT().IsLocalDevelopmentMode().Return(true)
				} else {
					env.EXPECT().IsLocalDevelopmentMode().Return(false)
				}
			}

			s := &statsd{
				env:          env,
				mdmsocketEnv: tt.mdmsocketstring,
			}

			network, address, err := s.getConnectionDetails()

			if err != nil && err.Error() != tt.wantError ||
				err == nil && tt.wantError != "" {
				t.Fatal(err)
			}

			if network != tt.network {
				t.Error(network)
			}
			if address != tt.address {
				t.Error(network)
			}
		})
	}
}

```
  
</details>